File: capy/src/e2e-cfun/java/dev/capylang/test/DeriveFeatureTest.java
package dev.capylang.test;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class DeriveFeatureTest {
    @Test
    void derivesShowForDataFields() {
        assertThat(DeriveFeature.showUser()).isEqualTo("DerivedUser { name, age }");
    }

    @Test
    void derivesShowForEmptyData() {
        assertThat(DeriveFeature.showEmpty()).isEqualTo("Empty {  }");
    }

    @Test
    void derivesShowFromLinkedInheritedFields() {
        assertThat(DeriveFeature.showEmployee()).isEqualTo("Employee { id, department }");
    }

    @Test
    void derivesShowForTypeFields() {
        assertThat(DeriveFeature.showNamed()).isEqualTo("Named { id }");
    }

    @Test
    void derivesMethodWithRuntimeParameter() {
        assertThat(DeriveFeature.userBiggerThan(40)).isTrue();
        assertThat(DeriveFeature.userBiggerThan(50)).isFalse();
    }

    @Test
    void derivesMethodWithMetadataPrimitiveAndDataParameters() {
        assertThat(DeriveFeature.userMixedParameterDeriveTrue()).isTrue();
        assertThat(DeriveFeature.userMixedParameterDeriveFalse()).isFalse();
    }

    @Test
    void derivesMethodWithReflectedFieldValues() {
        assertThat(DeriveFeature.userFieldValuesSummary()).isEqualTo("name=Ada,age=42");
    }
}

File: capy/src/e2e-cfun/js/dev/capylang/test/DeriveFeature.test.js
const test = require('node:test');
const { assert, generatedModule } = require('./_support');

test('DeriveFeature', () => {
    const derive = generatedModule('dev/capylang/test/DeriveFeature.js');
    assert.equal(derive.showUser(), 'DerivedUser { name, age }');
    assert.equal(derive.showEmpty(), 'Empty {  }');
    assert.equal(derive.showEmployee(), 'Employee { id, department }');
    assert.equal(derive.showNamed(), 'Named { id }');
    assert.equal(derive.userBiggerThan(40), true);
    assert.equal(derive.userMixedParameterDeriveFalse(), false);
    assert.equal(derive.userFieldValuesSummary(), 'name=Ada,age=42');
});
